ユーザー情報編集（管理者側）

Rihitonnnu opened the pull request:
## 今回のPRで行っていること
- ユーザー情報の編集と更新処理の実装
- 更新処理時のバリデーション作成
  - バリデーション条件は空欄が未記入の場合
- UI一部調整
  - ユーザー一覧画面
  - anchor-buttonコンポーネントの修正 

## UI変更点
### ユーザー情報編集画面
<img width="957" alt="image" src="https://user-images.githubusercontent.com/69156872/196613406-0222108b-a973-4587-aefc-d0568ece62a8.png">

### バリデーションがかかった時
<img width="945" alt="image" src="https://user-images.githubusercontent.com/69156872/196614324-87ab05ec-0b17-4cea-83cd-c346f0543d14.png">

### 修正後のユーザー一覧画面
少しカラムの横幅を修正しました。
<img width="941" alt="image" src="https://user-images.githubusercontent.com/69156872/196615870-fe6ad89d-914d-4688-8fc2-af49c2717ec3.png">

### 修正前のユーザー一覧画面
<img width="949" alt="193997250-931f93be-b5ef-494c-9277-ced14443ba50" src="https://user-images.githubusercontent.com/69156872/196616433-55d986ac-80da-4fce-b076-6fe134c8555f.png">

Rihitonnnu committed:
[update] aタグボタンコンポーネントのdivタグ削除

divタグを含むコンポーネントになっていてUIの調整がやりずらいためdivタグを削除しコンポーネントごとにdivタグを書くように修正しました。

Rihitonnnu committed:
[update] aタグボタンコンポーネントの修正に伴う変更とUI調整

[index.blade.php,show.blade.php]
aタグコンポーネントで削除したdivタグの追記を行いました。

[index.blade.php]
一覧ページUIの調整(paddingやwidth)を行いました。

Rihitonnnu committed:
[add] editメソッドとupdateメソッドの作成

Rihitonnnu committed:
[add] FormRequest(管理者側のユーザー情報更新)の作成

管理者側のユーザー情報更新処理にバリデーションをつけるため作成しました。

Rihitonnnu committed:
[add] ユーザー情報編集画面(管理者側)作成

YutakaOkabe commented on app/Http/Requests/Admin/User/UpdateRequest.php:
```diff
@@ -0,0 +1,39 @@
+    public function rules()
+    {
+        return [
+            'name' => 'required',
+            'email' => 'required',
```
どこまで厳密にやるかだけど、
nameはstring形式で文字数制限くらいはかけたい
emailは `required|email` でもいいし、もっとちゃんとやろうと思ったら `required|email:filter,dns` とかになるかな。少なくとも「@」あるよねくらいは確認したいな

YutakaOkabe commented on app/Http/Controllers/Admin/UserController.php:
```diff
@@ -24,4 +26,45 @@ public function show($id)
+        $user = User::find($id);
+
+        try {
+            DB::transaction(function () use ($user, $request) {
+                /** @var string $requestName */
+                $requestName = $request->name;
+
+                /** @var string $requestEmail */
+                $requestEmail = $request->email;
+
+                $user->name = $requestName;
+                $user->email = $requestEmail;
+
+                $user->save();
+            });
```
トランザクションは複数テーブルのデータをいっぺんに変更したい、かつどちらかのテーブルだけ更新されると困る時に使うので、今回は使わなくていいかなー

あとvalidation通った後の更新処理なのでもうちょっと簡潔に書いていいと思うよ！
何が入ってるか明示したいなら
```
$user->fill([
  'name'=>$request->input('name'),
  'email'=>$request->input('email'),
])->save();
```
でもいいし、ここは単純な更新だから
```
$user->fill($request)->save();
```
みたいな感じでも全然いいと思う

ryu022304 commented on app/Http/Controllers/Admin/UserController.php:
```diff
@@ -24,4 +26,45 @@ public function show($id)
+        /** @var \App\Models\User $user */
+        $user = User::find($id);
+
+        try {
+            DB::transaction(function () use ($user, $request) {
+                /** @var string $requestName */
+                $requestName = $request->name;
+
+                /** @var string $requestEmail */
+                $requestEmail = $request->email;
+
+                $user->name = $requestName;
+                $user->email = $requestEmail;
+
+                $user->save();
+            });
+        } catch (\Exception $e) {
+            var_dump($e->getMessage());
+        }
```
この辺りのDBの処理はベストプラクティス的な考えだと、Controllerじゃなくて、Modelに書いたほうが良いかなー
https://github.com/alexeymezenin/laravel-best-practices/blob/master/japanese.md#%E3%83%95%E3%82%A1%E3%83%83%E3%83%88%E3%83%A2%E3%83%87%E3%83%AB%E3%82%B9%E3%82%AD%E3%83%8B%E3%83%BC%E3%82%B3%E3%83%B3%E3%83%88%E3%83%AD%E3%83%BC%E3%83%A9
controllerの方はこんな感じに書けたらスッキリしそう。
```
if (!$user->update($request)) {
  // エラー処理
}
```
あと、複数のテーブルに対して更新処理を行っているわけではないので、わざわざトランザクション貼らなくても良い気はする。

YutakaOkabe commented on app/Http/Controllers/Admin/UserController.php:
```diff
@@ -24,4 +26,45 @@ public function show($id)
+        /** @var \App\Models\User $user */
+        $user = User::find($id);
+
+        try {
```
try~catchはちょっと冗長すぎるかな！
もうちょっと複雑でエラーで落とせない処理ならあっていいんだけど、今回は不要！

YutakaOkabe commented on app/Http/Controllers/Admin/UserController.php:
```diff
@@ -24,4 +26,45 @@ public function show($id)
+
+                $user->save();
+            });
+        } catch (\Exception $e) {
```
次どっかでtry-catch使う時に覚えてて欲しいんだけど、できるだけcatchするエラーは明確にしたい！
\Exception でcatchしたら全部のエラー拾えるから楽ではあるんだけど、意図しないエラーとかを気付けない原因になってしまうので！

YutakaOkabe commented on resources/views/admin/user/edit.blade.php:
```diff
@@ -0,0 +1,58 @@
+                    @if ($errors->any())
+                        <div class="alert alert-danger text-center">
+                            <ul>
+                                @foreach ($errors->all() as $error)
+                                    <li class="text-red-500">{{ $error }}</li>
+                                @endforeach
```
フォームの上にエラーを全部出す形式より、名前なら名前のフォームにエラーを出してあげるほうがユーザーには親切なので、個別にエラー出るように修正してみて！

YutakaOkabe reviewed (changes requested):
色々コメントしてる！わかんないところあったら気軽に聞いてくれー
あとテストも書きたい！あとで直接喋りに行くわ！

ryu022304 reviewed (commented):
見ましたー
貼ったベストプラクティスのページは見てみると結構勉強になると思う

Rihitonnnu committed:
[fix]ボタンが表示されなくなる部分の修正

サーバーを落とすたびにボタンが表示されない(表示はされているが変数で渡している背景色のクラスが適応されない(原因はまだわかっていない))のでコンポーネントごとに背景色を指定するclassをマージするように修正しました。

Rihitonnnu committed:
[update]バリデーション表示位置の修正

Rihitonnnu committed:
[update]メールアドレスのバリデーション修正

Rihitonnnu committed:
[clean]UserControllerのリファクタリング

メソッド内の記述をUserモデルに移動しました。

Rihitonnnu committed:
[clean]UserControllerのメソッド内容をUserモデルに移動

UserControllerではレスポンスのみを行うようにするためデータのやり取りをModel側へ移動しました。

Rihitonnnu committed:
[add]UserControllerのテスト作成

index～updateメソッドまでのテストを作成しました。

Rihitonnnu commented:
修正しましたのでレビューお願いします！

ボタンが表示されないバグのUIの画像を添付しておきます。
<img width="948" alt="Untitled" src="https://user-images.githubusercontent.com/69156872/197453255-6373cc1e-cdf8-4d07-aa18-033c09484466.png">

YutakaOkabe commented on resources/views/components/anchor-button.blade.php:
```diff
@@ -1,3 +1 @@
-    <a href="{{ $route }}" class="flex mx-auto text-white bg-{{ $color }}-500 border-0 py-2 px-6 focus:outline-none hover:bg-{{ $color }}-600 rounded">{{ $title }}</a>
-</div>
\ No newline at end of file
+<a href="{{ $route }}" {{$attributes->merge(['class'=>'flex mx-auto justify-center text-white border-0 py-2 px-6 focus:outline-none rounded'])}}>{{ $title }}</a>
```
ほーー、この書き方は汎用性高くて良いね 👍

YutakaOkabe commented on resources/views/admin/user/edit.blade.php:
```diff
@@ -0,0 +1,62 @@
+                                            <x-anchor-button route="{{ route('admin.user.index') }}" title="戻る" class="bg-red-500 hover:bg-red-600" />
+
+                                            {{-- 更新ボタン --}}
+                                            <x-anchor-button route="javascript:form1.submit()" title="更新する" class="bg-indigo-500 hover:bg-indigo-600" />
```
あーーなるほど、おれがリンクボタンはaタグで作ろうっていったからこっちも変えてくれたのね 😅
formの送信ボタンに関してはaタグで `javascript:form1.submit()` で送信するのはそんな見たことないので、普通にbuttonタグのtype=submitにしちゃってよいよ！
で他にもいっぱいsubmitボタン出てくるだろうから、submitボタンでひとつコンポーネントにしちゃっといた方がいいかも

Rihitonnnu replied:
勘違いですべてaタグで実装してました笑、追加でボタンタグのコンポーネントも作成しておきます！

YutakaOkabe commented on app/Http/Controllers/Admin/UserController.php:
```diff
@@ -21,7 +28,26 @@ public function index()
+     * @param int $id
+     *
+     * */
+    public function edit($id)
```
あ、ごめん今までのレビューで抜けてた 🙏 
引数に型を指定してあげましょう！自分が書くものに関してはできるだけ全部のメソッド！
```
public function edit(int $id)
```
こうしてあげると実行するときに引数に変な型が入ってきた段階で処理が落ちてくれるのでバグを発見しやすい！

Rihitonnnu replied:
了解です！修正しておきます！

YutakaOkabe commented on tests/Feature/Admin/User/UserControllerTest.php:
```diff
@@ -0,0 +1,70 @@
+
+class UserControllerTest extends TestCase
+{
+    /**
+     * A basic feature test example.
+     *
+     * @return void
+     */
```
ここのコメントはいらんな

ryu022304 replied:
`@test` ってアノテーションつけてあげるようにしても良いかもね。
そうすると関数名に `test` って付けなくても良くなる。
https://phpunit.readthedocs.io/ja/latest/annotations.html#test
テストの関数名どうするかは人やプロジェクトによるかなあと思うけれども、関数名に日本語入れているなら、 `test` とか付けないほうが統一感は出るかなという所感。

YutakaOkabe replied:
し、知らなかった 😅
ちなみにSRLは @test つけずに `test_日本語ほげほげ` で書いてますw 統一感的には日本語だけがいいですねw

YutakaOkabe commented on tests/Feature/Admin/User/UserControllerTest.php:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return void
+     */
+    public function test_ログインしていればユーザー一覧の表示()
+    {
+        /** @var \App\Models\Admin $admin*/
+        $admin = Admin::factory()->create();
+        $this->actingAs($admin, 'admins');
+
+        $response = $this->get(route('admin.user.index'));
+        $response->assertOk();
+
+        $response->assertViewIs('admin.user.index');
+    }
```
テストの書き方Good!
で欲を言えば別のテストメソッドでもう一個追加して欲しいのが、ユーザーがログインしていない場合にリダイレクトされるのも確認して欲しい！

Rihitonnnu replied:
うまくいっていない場合のテストが抜けていました！直しておきます！

YutakaOkabe commented on app/Http/Controllers/Admin/UserController.php:
```diff
@@ -21,7 +28,26 @@ public function index()
+     */
+    public function update(UpdateRequest $request, $id)
+    {
+        return to_route('admin.user.index', $this->user->updateUser($request, $id));
```
更新処理をModelに書いてくれてるのすごく良い！
その上で言うと、RequestクラスをそのままModelに渡してしまうと、ModelはHTTPのリクエストに依存してしまうことになるので、せっかく責任を分解しようとしてるのに依存関係が発生してしまうね
HTTPの処理はControllerに留めておきたいので、updateUserメソッドに渡すのはそれぞれのプロパティにしてしまいたい！なので書き方としては↓みたいな感じで書いてみてー
```
// controller
$this->user->updateUser($id, $request->name, $request->email);
// User
public function updateUser(int $id, string $name, string $name){}
```

Rihitonnnu replied:
確かにこのままだと表面上分けたように見えるだけですね...これから気をつけます！

YutakaOkabe commented on app/Models/User.php:
```diff
@@ -42,4 +42,16 @@ class User extends Authenticatable
+     * @return \App\Models\User
+     * @param int $id
+     */
+    public function updateUser($request, $id)
+    {
+        $user = $this::find($id);
+        $user->fill($request->all())->save();
+        return $user;
+    }
```
ControllerでRequestsクラスではなくプロパティをそれぞれこのメソッドに与えてあげることで、このメソッドは「飛んできた値を保存する」っていうことに集中できることになる
で、そんなふうに責任分解をするとこのクラスだけのテストが書けるようになるので、ぜひこの保存処理のテストも書いてみて！

Rihitonnnu replied:
了解です！リクエストとして受け取って保存ではなく、値として受け取って保存ですね！

YutakaOkabe commented on tests/Feature/Admin/User/UserControllerTest.php:
```diff
@@ -0,0 +1,70 @@
+
+        $user = User::factory()->create();
+
+        //編集画面
+        $response = $this->get(route('admin.user.edit', ['user' => $user->id]));
+        $response->assertOk();
+        $response->assertViewIs('admin.user.edit');
```
ここまででテストケースを切ってしまって、「ログインしていれば編集画面が表示される」とかのテストケースにしよう！
PHPUnitのassertは一つのテストが落ちるとそのテストケース内の他のassertを実行してくれなくなるので、できるだけ１つのテストケースは短くしたい！

Rihitonnnu replied:
なるほど、確かに全体を把握するためには区切ったほうがよさそうですね。ありがとうございます！

YutakaOkabe commented on tests/Feature/Admin/User/UserControllerTest.php:
```diff
@@ -0,0 +1,70 @@
+        $response->assertStatus(302); //リダイレクト
+        $response->assertSessionHasNoErrors();
+
+        //更新テスト用データで更新が正常に行われているか確認
+        $this->assertDatabaseHas('users', [
+            'name' => 'test',
+            'email' => 'test@gmail.com',
+        ]);
+    }
```
ここはModelの保存処理がうまくできてるかのテストなので、モデルのテストを作ったらそっちに移植してあげてよいよ
で、内容が変更されたかどうかはこいつの責任範囲ではなくなるので、このテストケースでやる内容は「updateDataをpostしたら302でindex画面に行く」だけでよい

Rihitonnnu replied:
ありがとうございます！責任範囲を意識できていなかったですね、モデルのテストを作成次第修正しておきます！

YutakaOkabe reviewed (changes requested):
ちゃんと理解して変更してくれてるのすごくよい！
たくさんコメントしちゃったけど、もうちょっとでかなり良くなると思うので頑張って！
わかんないことあったら聞いてくれー

Rihitonnnu committed:
[add]submitボタンコンポーネントの作成

formタグの送信を行うために用いるsubmitボタンのコンポーネントを作成しました。

Rihitonnnu committed:
[update]form送信を行うボタンをsubmitボタンコンポーネントに変更

Rihitonnnu committed:
[update]メソッドの引数の型指定追加

Rihitonnnu committed:
[update]UserControllerとUserの責任を分離

UserControllerのupdateメソッドからUserに対して、requestを渡すことによって発生する両者の依存関係を解決するために、requestのプロパティから値としてUserに渡すように修正しました。

Rihitonnnu committed:
[add]Userモデルのテスト作成

Rihitonnnu committed:
[update]不要なコメント文削除/testアノテーションの追加/メソッド名の日本語化

Rihitonnnu committed:
[update]データ更新のテストの分割/ログインしていない場合のリダイレクトのテスト作成

編集+更新処理をひとつのメソッドではなく２つのメソッドに分けてテストするように修正しました。
ログインしていない場合に管理者のログイン画面にリダイレクトされるかどうかを確認するテストを作成しました。

Rihitonnnu commented:
@YutakaOkabe 
修正しましたのでレビューお願いします！

ryu022304 commented on tests/Feature/Admin/User/UserTest.php:
```diff
@@ -0,0 +1,30 @@
+        //更新テスト用データで更新が正常に行われているか確認
+        $this->assertDatabaseHas('users', [
+            'name' => 'test',
+            'email' => 'test@gmail.com',
```
一つだけ問題見つけたのでコメント。
間違ってその人に送ってしまう可能性があるので、テスト用メールアドレスでgmailとかの普段使うドメインを設定するのは止めたほうが良いです。
自分が持っているメールアドレスとかなら良いけれども、今回だと `test@gmail.com` というメールアドレスに何かメールが飛んでしまう事故があり得るので。
なので、この辺にもまとめてあるけれども、結論、テスト用のドメインは `example.com` を使いましょうって感じですー。
https://qiita.com/suzutsuki0220/items/4ad83ed2e2adbb6507a4
今回メール飛ばす処理はないと思うけれども、普段から使うダミーデータは気をつけましょうってことで :hi:

Rihitonnnu replied:
確かにtest@gmail.comは使われている可能性は全然ありますもんね。もしメール送信機能あったら大変なことになってました…気をつけます！ありがとうございます！

YutakaOkabe commented on resources/views/components/submit-button.blade.php:
```diff
@@ -0,0 +1 @@
+<button type="submit" href="{{ $route }}" {{$attributes->merge(['class'=>'flex mx-auto justify-center text-white border-0 py-2 px-6 focus:outline-none rounded'])}}>{{ $title }}</a>
```
buttonにhref属性ってあったっけ？
formのactionでroute指定してあげてたらここでroute指定してあげなくても行くはず！試してみて！

Rihitonnnu replied:
確認してみたところ正常に動きました…修正しておきます！ありがとうございます。

YutakaOkabe commented on tests/Feature/Admin/User/UserControllerTest.php:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * @test
+     */
+    public function ログインしていなければ管理者のログイン画面へリダイレクトされる()
+    {
+        //一覧画面にログインなしでアクセスした場合
+        $response = $this->get(route('admin.user.index'));
+        $response->assertRedirect(route('admin.login'));
+
+        $user = User::factory()->create();
+        $response = $this->get(route('admin.user.show', ['user' => $user->id]));
+        $response->assertRedirect(route('admin.login'));
+
+        $response = $this->get(route('admin.user.edit', ['user' => $user->id]));
+        $response->assertRedirect(route('admin.login'));
+
+        //更新テスト用のデータ
+        $updateData = [
+            'name' => 'test',
+            'email' => 'test@gmail.com',
+        ];
+
+        $response = $this->put(route('admin.user.update', ['user' => $user->id]), $updateData);
+        $response->assertRedirect(route('admin.login'));
+    }
```
このテストケースも分割して、それぞれの正常系の動作の下に書いてしまいましょう！

Rihitonnnu replied:
了解しました！直しておきます！

YutakaOkabe commented on tests/Feature/Admin/User/UserTest.php:
```diff
@@ -0,0 +1,30 @@
+     */
+    public function データベースが更新されているかの確認()
+    {
+        //更新テスト用データで更新が正常に行われているか確認
```
あー、これだと結局Controllerのテストの実行結果に依存してしまうよね
1. このテストケースの中でuserをつくる
2. updateUserメソッドを呼ぶ
3. updateされてることを確認する

をこのテストケースの中でやりましょう！

Rihitonnnu replied:
ありがとうございます！確かにモデルの方ではなくコントローラーの方で更新してました…同じミス何回も申し訳ないです🙇‍♂️次は大丈夫だと思います！

YutakaOkabe reviewed (changes requested):
だいぶいい感じになってる！
テストケースのところでもうちょっとだけ修正して欲しいところがあるので、そこだけ綺麗になったら終わりだ！

Rihitonnnu committed:
[update]buttonに不要なhref属性の削除

Rihitonnnu committed:
[update]ダミーデータのメールアドレスを変更

ダミーデータのメールアドレスを実用的なドメインからexample.comに変更しました

Rihitonnnu committed:
[update]Userテストの修正

ユーザーデータの更新処理をUserControllerテストから分離させるため修正しました。

Rihitonnnu committed:
[update]未ログイン時のリダイレクトのテストの修正

まとめて書いていた未ログイン時にリダイレクトする処理のテストを分割しました。

Rihitonnnu commented:
@YutakaOkabe @ryu022304 
修正しましたのでレビューお願いします！

ryu022304 reviewed (approved):
![genba_cats_triple_check](https://user-images.githubusercontent.com/33801040/198192085-5f389453-2f54-4661-a333-49c9bb03e34b.jpg)

YutakaOkabe commented on tests/Feature/Admin/User/UserTest.php:
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Tests\Feature\Admin\User;
```
めっちゃ見逃してた
ModelのテストはUnitテストでやりましょう！UnitテストとFeatureテストの違いとかは調べてみて！
tests/Unit/Models ディレクトリを作ってあげて、そっちにModelのテストは書いていくようにしましょう！

Rihitonnnu replied:
ありがとうございます！Unitテストは粒度が細かく、Modelなどのテストで、Featureテストは粒度が大きめでControllerなどのテストでという感じで使い分けるんですね。知りませんでした...書き直しておきます。

YutakaOkabe commented on tests/Feature/Admin/User/UserTest.php:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * @test
+     */
+    public function モデルの存在確認()
+    {
+        $user = User::factory()->create();
+        $this->assertModelExists($user);
+    }
```
Unitテストに移行したらこのテストケースはいらないかな

YutakaOkabe commented on tests/Feature/Admin/User/UserTest.php:
```diff
@@ -0,0 +1,40 @@
+        $updatedUser = new User();
+        $updatedUser->updateUser($updateData['name'], $updateData['email'], $createdUser->id);
+
+        $this->assertDatabaseHas('users', [
+            'name' => 'test',
+            'email' => 'user@example.com',
+        ]);
```
今更だけど、このassertはあんまりよろしくないね
今ってテストケースごとにDBをリフレッシュしてないので（<- そもそもこれが良くない）、どこかのテストケースで作ったデータであってもデータが残ってしまっていて、ここでtrue判定されてしまってる可能性があると思う

Rihitonnnu replied:
なるほど。ではDBにデータを追加したり変更を加えたりするテストについては
`use RefreshDatabase;`を必ず書いておくべきなのでしょうか？

Rihitonnnu replied:
なるほど。ではDBにデータを追加したり変更を加えたりするテストについては
use RefreshDatabase;を必ず書いておくべきなのでしょうか？

YutakaOkabe replied:
そうね！DBを更新するテストケースでは必ず書いておきたい！というかDB更新しなくてもあって問題がないはずなので毎回書くのめんどいし、親の `Tests\TestCase` の方で書いちゃおうか！

Rihitonnnu replied:
了解です！ありがとうございます！

YutakaOkabe commented on tests/Feature/Admin/User/UserTest.php:
```diff
@@ -0,0 +1,40 @@
+            'email' => 'user@example.com',
+        ]);
+    }
+}
```
ここまで書いて、こんな感じでテスト書くんだよーって先に俺が見せてあげなかったのが非常に悪かったと反省した。
そりゃ手探りでやるのだいぶきついよね、、ごめんよーー🙇‍♂️🙇‍♂️

上に書いたようにModelのテストはUnitに移してしまって、中身としては↓ みたいな感じで書いてみて！
「一つのテストケースは他のテストケースと干渉しないように独立した形で書く」を意識してほしい！

```
<?php

namespace Tests\Unit\Models;

use App\Models\User;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;

class UserTest extends TestCase
{
    use RefreshDatabase;

    protected function setUp(): void
    {
        parent::setUp();
    }

    public function test_updateUser()
    {
        $user = User::factory()->create([
            'name' => 'hoge',
            'email' => 'hoge@example.com'
        ]);

        $updatedUser = (new User())->updateUser('fuga', 'fuga@example.com', $user->id);
        
        $this->assertEquals('fuga', $updatedUser->name);
        $this->assertEquals('fuga@example.com', $updatedUser->email);
    }
}
```

Rihitonnnu replied:
そんなことないです！自分はあれかなこれかなと調べながらやるのが好きなタイプなんで苦ではないです笑
ただテストの書き方はイマイチ腑に落ちないので教えて頂けるのはありがたいです🙇‍♂参考にさせていただきます！

YutakaOkabe reviewed (changes requested):
何度も差し戻してしまってごめんよ。。先に例を見せればよかったと非常に反省した🙇‍♂️
サンプルでコード書いてみたのでそれみながら書き直してもらえると！

Rihitonnnu committed:
UserTestをUnitディレクトリに移動/内容変更

Rihitonnnu committed:
TestcaseにRefreshDatabaseを追加

DBのリフレッシュを行うため追加しました

Rihitonnnu committed:
フィクスチャに管理者とユーザーのデータ作成を追加

Rihitonnnu committed:
UserControllerTestのフィクスチャ追加に伴う修正漏れの直し

Rihitonnnu commented:
@YutakaOkabe 
レビューお願いします！🙇‍♀

YutakaOkabe commented on app/Http/Controllers/Admin/UserController.php:
```diff
@@ -4,24 +4,54 @@
 
 class UserController extends Controller
 {
+    protected User $user;
+
+    public function __construct(User $user)
+    {
+        $this->user = $user;
+    }
```
こいつ使ってるっけ？消して問題なかったら消してて！

Rihitonnnu replied:
こいつはUser.phpのupdateUser()を呼び出す時に使ってますね。
```
   public function update(UpdateRequest $request, int $id)
    {
        /** @var string $userName */
        $userName = $request->name;
        /** @var string $userEmail */
        $userEmail = $request->email;
        return to_route('admin.user.index', $this->user->updateUser($userName, $userEmail, $id));
    }
```

YutakaOkabe replied:
あ、なるほど理解！

YutakaOkabe reviewed (approved):
1点コメントしてるので確認しといて！あとは良さげ、LGTM！お疲れ様〜

Rihitonnnu commented:
@ryu022304 @YutakaOkabe 
何度も細かくレビューして頂きありがとうございました🙇‍♀勉強になりました！
